### PR TITLE
Fix/redeem duplicated

### PR DESCRIPTION
**Pull request by wolfwarrier14**

### CODE REVIEW

File: logics/impls/controller/mod.rs [hunk 1]
```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
...
-            redeemer,
-            Some(pool),
-            redeem_amount,
-            0,
-            caller_pool,
-        )?;
-        if !shortfall.is_zero() {
-            return Err(Error::InsufficientLiquidity)
-        }
+        if let Some(oracle) = self._oracle() {
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:699]: If oracle.unwrap fails, it is just an Err (which basically should not happen), so I want to reduce nesting.
Can we make nesting unnecessary like this?
```rust
let oracle = self._oracle().ok_or(Error::OracleIsNotSet)?;
```

File: logics/impls/controller/mod.rs [hunk 2]
```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
...
+            let mut asset_price: u128 = 0;
+            let mut liquidation_threshold: u128 = 0;
+            let (shortfall, account_data) = {
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:702]: (Although either way of assigning the result of a calculation is fine.) Since we are using two different way, let's align them with one of them.

```rust
let (shortfall, account_data, asset_price, liquidation_threshold) = { ... }
```

File: logics/impls/controller/mod.rs [hunk 3]
```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
...
+                let mut avg_ltv = U256::from(0);
+                let mut avg_liquidation_threshold = U256::from(0);
+                let mut total_debt_in_base_currency: U256 = U256::from(0);
+
+                if let Some(pool_attribute) = pool_attributes {
+                    // if caller is a pool, get parameters for the pool without call the pool
+                    if pool_attribute.underlying.is_none() {
+                        return Err(Error::UnderlyingIsNotSet)
+                    }
+                    if pool_attribute.pool.is_none() {
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:720]: I think we are checking the same thing.

Comment (un-ast) [reply logics/impls/controller/mod.rs:720]: I think ok_or can be used since the unwrapped result is used in a subsequent step.
```rust
let underlying_in_attr =
    pool_attribute.underlying.ok_or(Error::UnderlyingIsNotSet)?;
```

File: logics/impls/controller/mod.rs [hunk 4]
```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
...
+                        return Err(Error::PoolIsNotSet)
+                    }
 
-        // FEATURE: update governance token supply index & distribute
+                    let collateral_factor_mantissa: Option<WrappedU256> =
+                        self.collateral_factor_mantissa(pool_attribute.pool.unwrap());
+                    if collateral_factor_mantissa.is_none() {
+                        return Err(Error::MarketNotListed)
+                    }
+                    let ltv = U256::from(collateral_factor_mantissa.unwrap());
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:732]:
```rust
let collateral_factor_mantissa = self
    .collateral_factor_mantissa(underlying_in_attr)
    .ok_or(Error::MarketNotListed)?;
let ltv = U256::from(collateral_factor_mantissa);
```

File: logics/impls/controller/mod.rs [hunk 5]
```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
...
+                        PriceOracleRef::get_price(&oracle, pool_attribute.underlying.unwrap());
+                    if let None | Some(0) = oracle_price {
+                        return Err(Error::PriceError)
+                    }
+                    let oracle_price_mantissa = Exp {
+                        mantissa: WrappedU256::from(U256::from(oracle_price.clone().unwrap())),
+                    };
+
+                    asset_price = oracle_price.unwrap();
+                    let compounded_liquidity_balance = pool_attribute.account_balance;
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:746]: It would be easier to see the following declarations considering the order of appearance

```rust
let compounded_liquidity_balance = pool_attribute.account_balance;
if compounded_liquidity_balance != 0 { ... }

let borrow_balance_stored = pool_attribute.account_borrow_balance;
if borrow_balance_stored != 0 { ... }
```

File: logics/impls/controller/mod.rs [hunk 6]
```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
...
+                        continue
+                    }
+                    // Read the balances and exchange rate from the pool
+                    let (
+                        compounded_liquidity_balance,
+                        borrow_balance_stored,
+                        exchange_rate_mantissa,
+                    ) = PoolRef::get_account_snapshot(&asset, redeemer);
+
+                    if compounded_liquidity_balance == 0 && borrow_balance_stored == 0 {
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:797]: You can also comment here about this

>I would like to see a comment in the code comments for future maintenance developers as to why _account_assets is not used in NOTE:.
>I understand that you are avoiding the Pool call on the _account_assets side and doing a similar check subsequent to this function.

File: logics/impls/controller/mod.rs [hunk 7]
```diff
@@ -1214,7 +1402,9 @@ impl<T: Storage<Data>> Internal for T {
         caller_pool: Option<(AccountId, PoolAttributes)>,
     ) -> Result<(U256, U256)> {
         // For each asset the account is in
-        let account_assets = self._account_assets(account, token_modify);
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:1217]: I would like to see a comment in the code comments for future maintenance developers as to why _account_assets is not used in NOTE:.
I understand that you are avoiding the Pool call on the _account_assets side and doing a similar check subsequent to this function.

File: logics/impls/controller/mod.rs [hunk 8]
```diff
@@ -1243,14 +1433,23 @@ impl<T: Storage<Data>> Internal for T {
...
-            for asset in &account_assets {
+            for asset in &markets {
+                if *asset == caller || *asset == skip_pool {
+                    continue
+                }
                 // Read the balances and exchange rate from the pool
                 let (token_balance, borrow_balance, exchange_rate_mantissa) =
                     PoolRef::get_account_snapshot(asset, account);
+
+                if token_balance == 0 && borrow_balance == 0 {
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:1449]: You can also comment here about this

>I would like to see a comment in the code comments for future maintenance developers as to why _account_assets is not used in NOTE:.
>I understand that you are avoiding the Pool call on the _account_assets side and doing a similar check subsequent to this function.

File: logics/impls/controller/mod.rs [hunk 9]
```diff
@@ -1214,7 +1402,9 @@ impl<T: Storage<Data>> Internal for T {
         caller_pool: Option<(AccountId, PoolAttributes)>,
     ) -> Result<(U256, U256)> {
         // For each asset the account is in
-        let account_assets = self._account_assets(account, token_modify);
+        let markets = self._markets();
+        let caller = Self::env().caller();
+        let mut skip_pool = caller;
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:1407]: How about declaring skip_pool as Option<AccountId> as follows?

```rust
let mut skip_pool = None;
```



Without caller_pool, both caller and skip_pool will contain Self::env().caller(), which makes the following branch a little difficult to understand

```rust
if *asset == caller || *asset == skip_pool {
    continue
}
```

File: logics/impls/controller/utils.rs [hunk 10]
```diff
@@ -140,6 +140,10 @@ pub fn get_hypothetical_account_liquidity(
         borrow_amount,
     } = input;
 
+    if asset_params.is_empty() {
+        sum_borrow_plus_effect = U256::from(redeem_tokens + borrow_amount);
+    }
```
Comment (un-ast) [inline logics/impls/controller/utils.rs:145]: Why is this necessary?
Is this different from what is considered in the following codes that follow?
```rust
if token_modify.is_some() && param.asset == token_modify.unwrap() { ... }
```

File: logics/impls/pool/mod.rs [hunk 11]
```diff
@@ -847,26 +849,31 @@ impl<T: Storage<Data> + Storage<psp22::Data> + Storage<psp22::extensions::metada
             let (_, account_borrow_balance, exchange_rate) = self.get_account_snapshot(redeemer);
             let account_balance = Internal::_balance_of(self, &redeemer);
             let contract_addr = Self::env().account_id();
-            let pool_attributes = PoolAttributesForWithdrawValidation {
+            // let pool_attributes = PoolAttributesForWithdrawValidation {
```
Comment (un-ast) [inline logics/impls/pool/mod.rs:852]: Were these comment-outs left intentionally?

File: logics/impls/controller/mod.rs [hunk 12]
```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
...
-        if !shortfall.is_zero() {
-            return Err(Error::InsufficientLiquidity)
-        }
+        if let Some(oracle) = self._oracle() {
+            let mut asset_price: u128 = 0;
+            let mut liquidation_threshold: u128 = 0;
+            let (shortfall, account_data) = {
+                // For each asset the account is in
+                let caller = Self::env().caller();
+                let mut skip_pool = caller;
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:705]: Same as comment in _get_hypothetical_account_liquidity.

**Review by un-ast — COMMENTED** [review]
Once we have reviewed the entire code
I haven't looked at the _redeem_allowed fine code yet

I think the _redeem_allowed change is to avoid making the same contract call in _get_hypothetical_account_liquidity and _calculate_user_account_data to collect the input part of the logic. Can't we just combine them (at some level) into a similar function and then write tests for that function? The purpose of the code is the same, but the code is not functionalized and written line by line, so I think the maintainability/readability is low (and reviewing is time-consuming).

File: logics/impls/controller/mod.rs (hunk 1 above)
Comment (wolfwarrier14) [reply logics/impls/controller/mod.rs]: I have resolved unnecessary nestings.

File: logics/impls/controller/mod.rs (hunk 3 above)
Comment (wolfwarrier14) [reply logics/impls/controller/mod.rs]: Done

File: logics/impls/controller/mod.rs (hunk 4 above)
Comment (wolfwarrier14) [reply logics/impls/controller/mod.rs]: Done

File: logics/impls/controller/utils.rs (hunk 10 above)
Comment (wolfwarrier14) [reply logics/impls/controller/utils.rs:145]: If asset_params is empty vector, sum_borrow is gettin 0 and not correct.

Comment (un-ast) [reply logics/impls/controller/utils.rs:145]: I understand what you are concerned about.
It just doesn't seem to be scaled or converted to the subsequent scale.

```rust
            // redeem effect
            // sumBorrowPlusEffects += tokensToDenom * redeemTokens
            sum_borrow_plus_effect = token_to_denom.clone().mul_scalar_truncate_add_uint(
                to_flatten(U256::from(redeem_tokens)),
                sum_borrow_plus_effect,
            );

            // borrow effect
            // sumBorrowPlusEffects += oraclePrice * borrowAmount
            sum_borrow_plus_effect = param
                .oracle_price_mantissa
                .clone()
                .mul_scalar_truncate_add_uint(
                    to_flatten(U256::from(borrow_amount)),
                    sum_borrow_plus_effect,
                );
```

Also, it seems that this function is currently only used in logics/impls/controller/mod#_get_hypothetical_account_liquidity
If there is a pattern where asset_params is empty but token_modify has a value, it needs to be fixed immediately.

File: logics/impls/controller/mod.rs [hunk 13]
```diff
@@ -718,24 +694,232 @@ impl<T: Storage<Data>> Internal for T {
...
         redeemer: AccountId,
         redeem_amount: Balance,
-        pool_attribute: Option<PoolAttributes>,
+        pool_attributes: Option<PoolAttributesForWithdrawValidation>,
     ) -> Result<()> {
-        let caller_pool = if let Some(_pool_attribute) = pool_attribute {
-            Some((pool, _pool_attribute))
-        } else {
-            None
+        let oracle = self._oracle().ok_or(Error::OracleIsNotSet)?;
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:699]: I understood that this function is written in the following flow.

```rust
    default fn _redeem_allowed(
        ...
    ) -> Result<()> {
        let oracle = self._oracle().ok_or(Error::OracleIsNotSet)?;
        let (shortfall, account_data, asset_price, liquidation_threshold) = {
            ...

            //// [1] Decomposition of pool_attribute
            if let Some(pool_attribute) = pool_attributes {
                ...
            }

            //// [2] Loop through all tokens on the market...
            //// Collect inputs for the following calculations
            ////// capacity calculation (※1)
            ////// health_factor calculation with liquidation_threshold (※2)
            for asset in markets {
                ...
            }

            //// [3] capacity calculation (※1)
            let (sum_collateral, sum_borrow_plus_effect) =
                get_hypothetical_account_liquidity(GetHypotheticalAccountLiquidityInput {
                    asset_params,
                    token_modify: Some(pool),
                    redeem_tokens: redeem_amount,
                    borrow_amount: 0,
                });

            ...

            //// [4] preliminary calculation/data collection as input for health_factor calculation with liquidation_threshold
            let health_factor = calculate_health_factor_from_balances(
                total_collateral_in_base_currency,
                total_debt_in_base_currency,
                avg_liquidation_threshold,
            );
            let account_data = AccountData {
                total_collateral_in_base_currency,
                total_debt_in_base_currency,
                avg_ltv,
                avg_liquidation_threshold,
                health_factor,
            };
            ...

            value
        };

        ...

        //// [5] health_factor calculation with liquidation_threshold (※2)
        let balance_decrease_allowed_result =
            balance_decrease_allowed(BalanceDecreaseAllowedParam {
                total_collateral_in_base_currency: account_data.total_collateral_in_base_currency,
                total_debt_in_base_currency,
                avg_liquidation_threshold: account_data.avg_liquidation_threshold,
                amount_in_base_currency_unit: U256::from(redeem_amount),
                asset_price: U256::from(asset_price),
                liquidation_threshold: U256::from(liquidation_threshold),
            });

        ...

        Ok(())
    }
```

Due to the large number of lines of code in this one function (over 200 lines), we have the following issues
- There are variables that appear only a little but are valid in scope for a long time
- There is a lot of code contained in one nest (important processes [1][2][3][4] are in one nest)

I would like to improve this in one of the following ways

**Option 1: organize nesting & order variable declarations & return values**

Separate the nests containing [1][2][3][4].
And the processes for capacity calculation and health_factor calculation with liquidation_threshold.

Here's the image.

```rust
    default fn _redeem_allowed(...) -> Result<()> {
        let oracle = self._oracle().ok_or(Error::OracleIsNotSet)?;
        let caller = Self::env().caller();

        let mut total_collateral_in_base_currency = U256::from(0);
        let mut avg_ltv = U256::from(0);
        let mut avg_liquidation_threshold = U256::from(0);
        let mut total_debt_in_base_currency = U256::from(0);

        //// [1] Decomposition of pool_attribute
        let (asset_price, liquidation_threshold, skip_pool) =
            if let Some(pool_attribute) = pool_attributes {
                // Collect input for capacity calculation
                asset_params.push(HypotheticalAccountLiquidityCalculationParam { ... });

                // Collect input for health_factor calculation with liquidation_threshold
                let liquidation_threshold = pool_attribute.liquidation_threshold;

                let compounded_liquidity_balance = pool_attribute.account_balance;
                if compounded_liquidity_balance != 0 { ... }

                let borrow_balance_stored = pool_attribute.account_borrow_balance;
                if borrow_balance_stored != 0 { ... }

                (oracle_price, liquidation_threshold, pool)
            } else {
                (0, 0, caller)
            };
        
        //// [2] Loop through all tokens on the market...
        for asset in markets {
          ...

          let (compounded_liquidity_balance, borrow_balance_stored, exchange_rate_mantissa) =
              PoolRef::get_account_snapshot(&asset, redeemer);
          //// NOTE: do not use account_asset function as it is making duplicated cross-contract calling leads to high gas.
          if compounded_liquidity_balance == 0 && borrow_balance_stored == 0 {
              continue
          }

          // Collect input for capacity calculation
          let decimals = ...;
          let oracle_price: u128 = ...;
          let collateral_factor_mantissa = ...;

          ...
          asset_params.push(HypotheticalAccountLiquidityCalculationParam { ... });

          // Collect input for health_factor calculation with liquidation_threshold
          let ltv = U256::from(collateral_factor_mantissa);
          let liquidation_threshold = PoolRef::liquidation_threshold(&asset);

          if compounded_liquidity_balance != 0 {
              let liquidity_balance_eth = ...;
              total_collateral_in_base_currency = ...;
              avg_ltv = ...;
              avg_liquidation_threshold = ...;
          }

          if borrow_balance_stored != 0 {
              let borrow_balance_eth = ...
              total_debt_in_base_currency = ...;
          }
      }

      //// [3] capacity calculation (※1)
      let (sum_collateral, sum_borrow_plus_effect) =
          get_hypothetical_account_liquidity(GetHypotheticalAccountLiquidityInput { ... });
      let shortfall = if sum_collateral > sum_borrow_plus_effect {
          U256::from(0)
      } else {
          sum_borrow_plus_effect.sub(sum_collateral)
      };
      if !shortfall.is_zero() {
          return Err(Error::InsufficientLiquidity)
      }

      //// [4] preliminary calculation/data collection as input for health_factor calculation with liquidation_threshold
      avg_ltv = if total_collateral_in_base_currency.is_zero() {
          U256::from(0)
      } else {
          avg_ltv.div(total_collateral_in_base_currency)
      };
      avg_liquidation_threshold = if total_collateral_in_base_currency.is_zero() {
          U256::from(0)
      } else {
          avg_liquidation_threshold.div(total_collateral_in_base_currency)
      };
      let health_factor = calculate_health_factor_from_balances( ... );
      let account_data = AccountData { ... };

      //// [5] health_factor calculation with liquidation_threshold (※2)
      let total_debt_in_base_currency = account_data.total_debt_in_base_currency;
      if total_debt_in_base_currency.is_zero() {
          return Ok(())
      }
      let balance_decrease_allowed_result =
          balance_decrease_allowed(BalanceDecreaseAllowedParam { ... });
      if !balance_decrease_allowed_result {
          return Err(Error::BalanceDecreaseNotAllowed)
      }

      ...
    }
```

Nesting involving [1][2][3][4] as shown above is gone,
I think it will reduce the number of updatable variables (except for arg_xxx at the beginning) and make it easier to see where the variables are for.

**Option 2: Write a test that cuts the process into a function and shares it with existing logic**

Option1 may be better for code brevity, but may be preferable for functional quality.

Most of the code here is
- Check for surplus capacity (flow up to call _get_hypothetical_account_liquidity in Controller#_borrow_allowed)
- Controller#__balance_decrease_allowed
and reordered to reduce the number of Pool calls, etc.

Therefore, we should check the quality of each location function (ex: redeem, transfer...) at the same time by creating functions that share code with these existing logics and testing them.

**Review by un-ast — COMMENTED** [review]
@wolfwarrier14 
Concerned about the amount of code in redeem_allowed and the duplication of existing code, we proposed an improvement. Please take a look and let me know which way you would like to address this.

File: logics/impls/controller/mod.rs (hunk 13 above)
Comment (wolfwarrier14) [reply logics/impls/controller/mod.rs:717]: Do you want me to separate the contents into functions?

Comment (un-ast) [reply logics/impls/controller/mod.rs:717]: it would be better if the code could be separated into functions (Option 2) for testability, but since the biggest problem is low readability due to nest and variable scopes.
It would be fine to maintain this function in the manner of Option 1, up to and including rearrangement of the description order and scope.

File: logics/impls/controller/mod.rs [hunk 14]
```diff
@@ -1516,83 +1293,119 @@ impl<T: Storage<Data>> Internal for T {
...
+            let oracle_price_mantissa = Exp {
+                mantissa: WrappedU256::from(U256::from(oracle_price)),
+            };
+
+            let collateral_factor_mantissa = self
                 ._collateral_factor_mantissa(asset)
                 .ok_or(Error::MarketNotListed)?;
             let ltv = U256::from(collateral_factor_mantissa);
 
             let liquidation_threshold = PoolRef::liquidation_threshold(&asset);
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:1405]: [WANT]
For gas reduction, it may be a good idea to put together a function to get the metadata of a pool.
(Further optimization will increase the amount of changes, so this seems to be all we need for now.)

```rust
PoolRef::token_decimals(&asset)
PoolRef::liquidation_threshold(&asset);
```

Comment (un-ast) [reply logics/impls/controller/mod.rs:1405]: I don't want to put it in the scope right now,  but in the future, I think it would be good to organize it into a function that can get all the information from the Pool at once, since I think there will be a lot of gas in the Pool calls
For example
[1] PoolRef::metadata
Get pool metadata (underlying, decimals, liquidation_threshold etc.)
[2] PoolRef::status
get (potentially variable) status of the pool (total supply, exchange rate, etc.)
[3] PoolRef::account_snapshot
information per account (≈ get_account_snapshot)

**Review by un-ast — COMMENTED** [review]
@wolfwarrier14 
Looks good. I put in my comments where I could comment first on the larger portion. Now I will look at the details of _redeem_allowed.

File: logics/impls/controller/mod.rs [hunk 15]
```diff
@@ -267,6 +266,7 @@ pub trait Internal {
         &self,
         account: AccountId,
         pool_attributes: Option<PoolAttributes>,
+        get_asset_param: bool,
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:269]: Please tell us why you want to include this flag.

I don't think this flag is necessary.
What this flag does is only push to Vec or not, and the amount of processing (especially by cross contract call) is almost the same, and I think it is confusing because Vec<HypotheticalAccount LiquidityCalculationParam> exists (empty or not) in the return value, so I think the flag makes it more difficult.

**Review by un-ast — COMMENTED** [review]
@wolfwarrier14 I commented on the added commitments

File: logics/impls/controller/mod.rs (hunk 15 above)
Comment (wolfwarrier14) [reply logics/impls/controller/mod.rs:269]: I thought that reducing vector operation can reduce gas.
But it the gas cost is similar this flag doesn't needed.
I will remove it.

Comment (un-ast) [reply logics/impls/controller/mod.rs]: I think it will certainly decrease slightly.
However, I suspect that many of them are cross contract calls, and if we have this one, we would also want a flag for AccountCollateralData, which would also complicate things.

Comment (wolfwarrier14) [reply logics/impls/controller/mod.rs]: Yes, I agree,
I have removed the flag.

File: logics/impls/controller/mod.rs [hunk 16]
```diff
@@ -713,33 +719,69 @@ impl<T: Storage<Data>> Internal for T {
...
+                redeem_tokens: redeem_amount,
+                borrow_amount: 0,
+            });
+
+        // These are safe, as the underflow condition is checked first
+        let shortfall = if sum_collateral > sum_borrow_plus_effect {
+            U256::from(0)
         } else {
-            None
+            sum_borrow_plus_effect.sub(sum_collateral)
```
Comment (un-ast) [inline logics/impls/controller/mod.rs:756]: I think I can summarize since the only code that uses shortfall is the one immediately after

```rust
        if sum_collateral < sum_borrow_plus_effect {
            return Err(Error::InsufficientLiquidity)
        }
```

**Review by un-ast — COMMENTED** [review]
@wolfwarrier14 
I reviewed around redeem_allowed, looks good!
Could you please fix it?

File: logics/impls/controller/mod.rs [hunk 17]
```diff
@@ -1211,296 +1266,291 @@ impl<T: Storage<Data>> Internal for T {
...
+        let (asset_price, liquidation_threshold, skip_pool) = if let Some(pool_attribute) =
+            pool_attributes
+        {
+            // if caller is a pool, get parameters for the pool without call the pool
+            let attr_underlying = pool_attribute.underlying.ok_or(Error::UnderlyingIsNotSet)?;
+            let attr_pool = pool_attribute.pool.ok_or(Error::PoolIsNotSet)?;
+
+            let collateral_factor_mantissa = self
+                ._collateral_factor_mantissa(attr_pool)
+                .ok_or(Error::MarketNotListed)?;
```
[the remark this answers is not in the document]
Comment (wolfwarrier14) [reply logics/impls/controller/mod.rs:1318]: Yeah, it would be better to use InvalidCollateralFactor to understand easily.

File: logics/impls/pool/mod.rs [hunk 18]
```diff
@@ -1378,37 +1358,34 @@ impl<T: Storage<Data> + Storage<psp22::Data> + Storage<psp22::extensions::metada
...
-
-            let balance_decrease_allowed = ControllerRef::balance_decrease_allowed(
-                &controller,
-                pool_attributes,
-                user,
-                account_balance,
-            )?;
-
-            if !balance_decrease_allowed {
-                return Err(Error::DepositAlreadyInUse)
```
Comment (un-ast) [inline logics/impls/pool/mod.rs:1405]: "DepositAlreadyInUse" one no longer appears to be in use.

File: .github/workflows/ci.yml [hunk 19]
```diff
@@ -137,24 +137,24 @@ jobs:
...
+      #       primes-${{ runner.os }}-${{ runner.arch }}-
+      # - name: Restore cached Artifacts
+      #   id: cache-artifacts-restore
+      #   uses: actions/cache/restore@v3
+      #   with:
+      #     path: |
+      #       **/artifacts
+      #     key: artifacts-${{ runner.os }}-${{ runner.arch }}-${{ github.sha }}
+      #     restore-keys: |
+      #       artifacts-${{ runner.os }}-${{ runner.arch }}-
```
Comment (un-ast) [inline .github/workflows/ci.yml:157]: Isn't this commented out for testing purposes?

File: .github/workflows/ci.yml [hunk 20]
```diff
@@ -179,20 +179,20 @@ jobs:
...
+      #     path: |
+      #       **/target
+      #     key: primes-${{ runner.os }}-${{ runner.arch }}-${{ github.sha }}
+      # - name: Save Artifacts
+      #   id: cache-artifacts-save
+      #   uses: actions/cache/save@v3
+      #   with:
+      #     path: |
+      #       **/artifacts
+      #     key: artifacts-${{ runner.os }}-${{ runner.arch }}-${{ github.sha }}
```
Comment (un-ast) [inline .github/workflows/ci.yml:195]: Isn't this commented out for testing purposes?

File: tests/Controller.spec.ts [hunk 21]
```diff
@@ -765,7 +765,7 @@ describe('Controller spec', () => {
             )
           ).value.ok.ok,
           {
-            collateral: 90 - (50 * 90) / 100,
+            collateral: 90,
```
Comment (un-ast) [inline tests/Controller.spec.ts:768]: This fix should be a code optimization, not a feature change, but do you see what the difference is here?

File: tests/Controller.spec.ts [hunk 22]
```diff
@@ -780,8 +780,8 @@ describe('Controller spec', () => {
             )
           ).value.ok.ok,
           {
-            collateral: 0,
-            shortfall: 500 - 450,
+            collateral: 450,
+            shortfall: 0,
```
Comment (un-ast) [inline tests/Controller.spec.ts:784]: Same
> This fix should be a code optimization, not a feature change, but do you see what the difference is here?

Furthermore, the state of the capacity is reversed here, which is of particular concern.

**Review by un-ast — COMMENTED** [review]
@wolfwarrier14 I thought the code was good, but I am concerned about the status of the test.

File: .github/workflows/ci.yml (hunk 19 above)
Comment (wolfwarrier14) [reply .github/workflows/ci.yml:157]: This is the cash operation and ci is using too much cache and can cause the disk lack error.

File: tests/Controller.spec.ts (hunk 21 above)
Comment (wolfwarrier14) [reply tests/Controller.spec.ts:768]: Yeah, there was an logic error in controller/utils.rs `get_hypothetical_account_liquidity` function

File: tests/Lens.spec.ts [hunk 23]
```diff
@@ -67,17 +66,11 @@ const setup = async (
     args: [[ONE_ETHER], [ONE_ETHER], [ONE_ETHER], [ONE_ETHER]],
   })
 
-  const incentivesController = await deployIncentivesController({
-    api,
-    signer: deployer,
-    args: [],
-  })
-
```
Comment (un-ast) [inline tests/Lens.spec.ts:75]: Why should this and the subsequent following be deleted?

```diff
    args: [
-      incentivesController.address,
+      null,
```

Comment (wolfwarrier14) [reply tests/Lens.spec.ts:75]: incentiveController is optional and it doesn't affect.

Comment (wolfwarrier14) [reply tests/Lens.spec.ts:75]: I just tried to test if removing incentives controller can remove timeout error in e2e tests.

File: .github/workflows/ci.yml (hunk 19 above)
Comment (un-ast) [reply .github/workflows/ci.yml:157]: I see. Can I get a link to the CI log that is failing with that error for reference?

Comment (wolfwarrier14) [reply .github/workflows/ci.yml:157]: https://github.com/starlay-finance/starlay-protocol-wasm/actions
I couldn't find the detailed link but there were some errors because of disk space.

Comment (wolfwarrier14) [reply .github/workflows/ci.yml:157]: https://github.com/starlay-finance/starlay-protocol-wasm/actions/runs/6428142442/job/17454864467

Comment (un-ast) [reply .github/workflows/ci.yml:157]: Thank you, this has nothing to do with this PR, so let's leave it at that for once.

I just don't know if that is the cause.

"No space left on device" is output with warn when saving

>Run actions/cache/save@v3
/usr/bin/tar --posix -cf cache.tzst --exclude cache.tzst -P -C /home/runner/work/starlay-protocol-wasm/starlay-protocol-wasm --files-from manifest.txt --use-compress-program zstdmt
zstd: error 70 : Write error : cannot write block : No space left on device 
/usr/bin/tar: cache.tzst: Wrote only 4096 of [1](https://github.com/starlay-finance/starlay-protocol-wasm/actions/runs/6428142442/job/17454864467#step:15:1)0240 bytes
/usr/bin/tar: Child returned status [7](https://github.com/starlay-finance/starlay-protocol-wasm/actions/runs/6428142442/job/17454864467#step:15:8)0
/usr/bin/tar: Error is not recoverable: exiting now
Warning: Failed to save: "/usr/bin/tar" failed with error: The process '/usr/bin/tar' failed with exit code 2
Warning: Cache save failed.

And I am getting an error message that there are no libraries needed.

>Run substrate-contracts-node --version
/home/runner/work/_temp/fc29[4](https://github.com/starlay-finance/starlay-protocol-wasm/actions/runs/6428142442/job/17454864467#step:17:5)[5](https://github.com/starlay-finance/starlay-protocol-wasm/actions/runs/6428142442/job/17454864467#step:17:6)23-3e3c-4284-929d-7c90b9981ff8.sh: line 1: substrate-contracts-node: command not found
Error: Process completed with exit code 127.

**Review by un-ast — APPROVED** [review]
